File: python/orderbook.py
from globals import satisfied_history

class OrderBook:

    def __init__(self):
        self.buy_head = None    # Linked list of buy orders in descending order
        self.sell_head = None   # Linked list of sell orders in ascending order

    def insert_order(self, order):
        if order.order_type == "BUY":
            # Insert into buy list in descending order
            if self.buy_head is None or order.price > self.buy_head.price:
                order.next = self.buy_head
                self.buy_head = order
            else:
                current = self.buy_head
                while current.next is not None and current.next.price >= order.price:
                    current = current.next
                order.next = current.next
                current.next = order
        elif order.order_type == "SELL":
            # Insert into sell list in ascending order
            if self.sell_head is None or order.price < self.sell_head.price:
                order.next = self.sell_head
                self.sell_head = order
            else:
                current = self.sell_head
                while current.next is not None and current.next.price <= order.price:
                    current = current.next
                order.next = current.next
                current.next = order
    
    def match_order(self):
        while (self.buy_head is not None and self.sell_head is not None and self.buy_head.price >= self.sell_head.price):

            match_price = self.sell_head.price    # Price at which order is matched

            match_quantity = min(self.buy_head.quantity, self.sell_head.quantity)   # Quantity of shares matched
            
            print(f"[Ticker {self.buy_head.ticker}] Order matched: {match_quantity} shares at ${match_price}")

            self.buy_head.quantity -= match_quantity
            self.sell_head.quantity -= match_quantity

            if self.buy_head.quantity <= 0:
                satisfied_history.append(f"BUY order satisfied: [Ticker {self.buy_head.ticker}] {self.buy_head.order_type} {match_quantity} @ ${self.buy_head.price}")
                self.buy_head = self.buy_head.next
            if self.sell_head is not None and self.sell_head.quantity <= 0:
                satisfied_history.append(f"SELL order satisfied: [Ticker {self.sell_head.ticker}] {self.sell_head.order_type} {match_quantity} @ ${self.sell_head.price}")
                self.sell_head = self.sell_head.next

File: java/StockEngine/src/OrderBook.java
import java.util.concurrent.atomic.AtomicReference;

public class OrderBook {

    private final AtomicReference<Order> buyHead;
    private final AtomicReference<Order> sellHead;

    public OrderBook() {
        this.buyHead = new AtomicReference<>(null);     // Linked list of BUY orders in descending order
        this.sellHead = new AtomicReference<>(null);    // Linked list of SELL orders in descending order
    }

    public AtomicReference<Order> getBuyHead() {
        return this.buyHead;
    }

    public AtomicReference<Order> getSellHead() {
        return this.sellHead;
    }

    public void insertOrder(Order order) {
        if (order.orderType == Order.OrderType.BUY) {
            insertBuyOrder(order);
        }
        else if (order.orderType == Order.OrderType.SELL) {
            insertSellOrder(order);
        }
    }

    private void insertBuyOrder(Order order) {
        while (true) {
            Order head = buyHead.get();
            // If first order in list
            if (head == null || order.price > head.price) {
                order.next.set(head);
                if (buyHead.compareAndSet(head, order)) {
                    return;
                }
            }
            // Normal insert in descending order
            else {
                Order pred = head;
                Order curr = pred.next.get();
                while (curr != null && curr.price >= order.price) {
                    pred = curr;
                    curr = curr.next.get();
                }
                order.next.set(curr);
                if (pred.next.compareAndSet(curr, order)) {
                    return;
                }
            }
        }
    }

    private void insertSellOrder(Order order) {
        while (true) {
            Order head = sellHead.get();
            // If first order in list
            if (head == null || order.price < head.price) {
                order.next.set(head);
                if (sellHead.compareAndSet(head, order)) {
                    return;
                }
            }
            // Normal insert in ascending order
            else {
                Order pred = head;
                Order curr = pred.next.get();
                while (curr != null && curr.price <= order.price) {
                    pred = curr;
                    curr = curr.next.get();
                }
                order.next.set(curr);
                if (pred.next.compareAndSet(curr, order)) {
                    return;
                }
            }
        }
    }

    public void matchOrder() {

        while (true) {

            Order buy = buyHead.get();
            Order sell = sellHead.get();
            if (buy == null || sell == null || buy.price < sell.price) {
                break;
            }

            int buyQty, sellQty, matchQty;
            while (true) {
                buyQty = buy.quantity.get();
                sellQty = sell.quantity.get();
                if (buyQty <= 0 || sellQty <= 0) {
                    break;
                }
                matchQty = Math.min(buyQty, sellQty);     // Price at which order is matched
                if (buy.quantity.compareAndSet(buyQty, buyQty - matchQty)) {    // Atomic function compareAndSet used to avoid interruption from other threads
                    if (sell.quantity.compareAndSet(sellQty, sellQty - matchQty)) {
                        System.out.println("[Ticker " + buy.ticker + "] Order matched: " + matchQty + " shares at $" + sell.price);
                        break;
                    } else {
                        continue;
                    }
                }
            }

            // If all BUY order shares are satisfied
            if (buy.quantity.get() == 0) {
                buyHead.compareAndSet(buy, (buy.next.get()));
            }
            // If all SELL order shares are satisfied
            if (sell.quantity.get() == 0) {
                sellHead.compareAndSet(sell, (sell.next.get()));
            }
        }
    }

    
}
